mate-session-manager-1.22.0 FTBFS

#### Expected behaviour


#### Actual behaviour


#### Steps to reproduce the behaviour


#### MATE general version


#### Package version


#### Linux Distribution


#### Link to downstream report of your Distribution

As per $SUBJECT, I get the following error when building mate-session-manager on a systemd-less system. The error reads:

gsm-systemd.c:432: error: undefined reference to 'sd_pid_get_session'
gsm-systemd.c:438: error: undefined reference to 'sd_uid_get_sessions'
gsm-systemd.c:453: error: undefined reference to 'sd_session_get_state'
gsm-systemd.c:464: error: undefined reference to 'sd_session_get_type'
collect2: error: ld returned 1 exit status

The following patch allows the build to complete, but is only lightly tested.

--- mate-session-manager-1.22.0/mate-session/gsm-systemd.c.orig 2019-03-04 15:14:01.668664101 +0000
+++ mate-session-manager-1.22.0/mate-session/gsm-systemd.c      2019-03-04 15:37:29.260698826 +0000
@@ -429,14 +429,18 @@ gsm_systemd_is_last_session_for_user (Gs
         gboolean is_last_session;
         int ret, i;
 
+#ifdef HAVE_SYSTEMD
         ret = sd_pid_get_session (getpid (), &session);
-
+#else
+       ret = -1;
+#endif
         if (ret != 0) {
                 return FALSE;
         }
 
+#ifdef HAVE_SYSTEMD
         ret = sd_uid_get_sessions (getuid (), FALSE, &sessions);
-
+#endif
         if (ret <= 0) {
                 free (session);
                 return FALSE;
@@ -450,8 +454,9 @@ gsm_systemd_is_last_session_for_user (Gs
                 if (g_strcmp0 (sessions[i], session) == 0)
                         continue;
 
+#ifdef HAVE_SYSTEMD
                 ret = sd_session_get_state (sessions[i], &state);
-
+#endif
                 if (ret != 0)
                         continue;
 
@@ -461,8 +466,9 @@ gsm_systemd_is_last_session_for_user (Gs
                 }
                 free (state);
 
+#ifdef HAVE_SYSTEMD
                 ret = sd_session_get_type (sessions[i], &type);
-
+#endif
                 if (ret != 0)
                         continue;
 
------
Hope this helps.

On a systemd free system I get:

gsm-systemd.c:432: error: undefined reference to 'sd_pid_get_session'
gsm-systemd.c:438: error: undefined reference to 'sd_uid_get_sessions'
gsm-systemd.c:453: error: undefined reference to 'sd_session_get_state'
gsm-systemd.c:464: error: undefined reference to 'sd_session_get_type'
collect2: error: ld returned 1 exit status

--- a/mate-session/gsm-systemd.c
+++ b/mate-session/gsm-systemd.c
@@ -429,16 +429,31 @@ gsm_systemd_is_last_session_for_user (GsmSystemd *manager)
         gboolean is_last_session;
         int ret, i;
 
+#ifdef HAVE_SYSTEMD
         ret = sd_pid_get_session (getpid (), &session);
+#endif
 
-        if (ret != 0) {
+        if (session == NULL) {
                 return FALSE;
         }
 
+        if (ret != 0) {
+                free (session);
+                return FALSE;
+        }
+
+#ifdef HAVE_SYSTEMD
         ret = sd_uid_get_sessions (getuid (), FALSE, &sessions);
+#endif
+
+        if (sessions == NULL) {
+                free (session);
+                return FALSE;
+        }
 
         if (ret <= 0) {
                 free (session);
+                free (sessions);
                 return FALSE;
         }
 
@@ -450,7 +465,9 @@ gsm_systemd_is_last_session_for_user (GsmSystemd *manager)
                 if (g_strcmp0 (sessions[i], session) == 0)
                         continue;
 
+#ifdef HAVE_SYSTEMD
                 ret = sd_session_get_state (sessions[i], &state);
+#endif
 
                 if (ret != 0)
                         continue;
@@ -461,7 +478,9 @@ gsm_systemd_is_last_session_for_user (GsmSystemd *manager)
                 }
                 free (state);
 
+#ifdef HAVE_SYSTEMD
                 ret = sd_session_get_type (sessions[i], &type);
+#endif
 
                 if (ret != 0)
                         continue;

